content/blog/2025-03-21-parquet-pushdown.md

```diff
@@ -0,0 +1,312 @@
+
+2. Decompressed pages are cached for minimal time; after one pass (steps 1-6), the cache memory is released for the next pass. 
+
+This allows the cache to only hold 1 page at a time, and to immediately discard the previous page after it's used, significantly reducing the memory requirement for caching.
+
+### What pages are cached?
+You may have noticed that only `location` is cached, not `val`, because `val` is only used for output.
+More generally, only columns that appear both in the filter and output are cached, and at most 1 page is cached for each such column.
+
+More examples:
+```sql
+SELECT val 
+FROM sensor_data 
+WHERE date_time > '2025-03-11' AND location = 'office';
+```
+
+In this case, no columns are cached, because `val` is not used for filtering.
+
+```sql
+SELECT COUNT(*) 
+FROM sensor_data 
+WHERE date_time > '2025-03-11' AND location = 'office';
+```
+
+In this case, again, no columns are cached, because the output projection is empty after query plan optimization.
+
+### Then why cache 2 pages per column instead of 1? 
+This is another real-world nuance regarding how Parquet layouts the pages.
+
+Parquet by default encodes data using [dictionary encoding](https://parquet.apache.org/docs/file-format/data-pages/encodings/), which writes a dictionary page as the first page of a column chunk, followed by the keys referencing the dictionary.
+
+You can see this in action using [parquet-viewer](https://parquet-viewer.xiangpeng.systems):
+
+<figure>
+  <img src="/blog/images/parquet-pushdown/parquet-viewer.jpg" alt="Parquet viewer shows the page layout of a column chunk" width="80%" class="img-responsive">
+  <figcaption>
+    Parquet viewer shows the page layout of a column chunk
+  </figcaption>
+</figure>
+
+This means that to decode a page of data, it actually references two pages: the dictionary page and the data page.
+
+This is why it caches 2 pages per column: one dictionary page and one data page.
+The data page slot will move forward as it reads the data; but the dictionary page slot always references the first page.
+
+<figure>
+  <img src="/blog/images/parquet-pushdown/cached-pages.jpg" alt="Cached two pages, one for dictionary (pinned), one for data (moves as it reads the data)" width="80%" class="img-responsive">
+  <figcaption>
+    Cached two pages, one for dictionary (pinned), one for data (moves as it reads the data)
+  </figcaption>
+</figure>
+
+
+## How does it perform?
+
+Here are my results on [ClickBench](https://github.com/apache/datafusion/tree/main/benchmarks#clickbench) on my AMD 9900X machine. The total time is reduced by 15%, with Q23 being 2.24x faster,
+and queries that get slower are likely due to noise.
+
+```
+┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
+┃ Query        ┃ no-pushdown ┃ new-pushdown ┃        Change ┃
+┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
+│ QQuery 0     │      0.47ms │       0.43ms │ +1.10x faster │
```

Comment:
Any thoughts on using percent to communicate difference in performance. 
`+1.10x faster` is not immediately obvious is the same as `10% faster`.

Reply:
I used this script to generate the table: https://github.com/apache/datafusion/blob/main/benchmarks/compare.py

I agree the exact presentation can be improved, but it is consistent with other datafusion benchmark reports.